### PR TITLE
Enable NetworkInformation tests on Linux, OSX

### PR DESCRIPTION
**Pull request by mellinoe**
cc: @stephentoub , @CIPop 

This removes the "Unsupported" tag for OSX and Linux from the NetworkInformation tests. Some of the tests just attempt to enumerate all of the properties and statistics that the library exposes. I've disabled those test cases on Linux and OSX because many of those members are not supported.

I encountered a couple small issues after running the tests and fixed the issues in subsequent commits:

* In Linux's NetworkChange implementation, the field storing the netlink socket's file descriptor was not being set to 0 after it was closed. This means if you tried to close the socket, and then re-open or re-close it, you'd hit a debug assertion I had in place. I've set the field to 0 after it is closed, and clarified the debug assertion messages.

* UnixIP(v4/v6)InterfaceProperties was caching the network interface's index too early, and was always storing 0. The network interface's index isn't initialized until the link-layer address is enumerated from the native shim call. To avoid timing issues like this, I might consider storing all of the information from the address enumeration in a separate object, and then constructing the UnixNetworkInterface object all at once.

* OSX didn't implement OperationalStatus. I've added a very crude implementation which just determines whether an interface is "Up" or not by checking if it has any addresses assigned to it. I think the OS exposes a bit more information here, but I don't think that we can necessarily map that info to our enum values (except perhaps "Dormant").

### CODE REVIEW

File: src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Linux.cs [hunk 1]
```diff
@@ -35,7 +35,7 @@ public class NetworkChange
                 {
                     if (s_addressChangedSubscribers == null)
                     {
-                        Debug.Assert(s_socket == 0);
+                        Debug.Assert(s_socket == 0, "s_socket != 0, but there are no subscribers to NetworkAddressChanged.");
                         return;
                     }
```
Comment (stephentoub) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Linux.cs]: Should we also assert here that s_socket != 0?

File: src/System.Net.NetworkInformation/tests/FunctionalTests/IPInterfacePropertiesTest.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@ public IPInterfacePropertiesTest()
         }
 
         [Fact]
+        [PlatformSpecific(PlatformID.Windows)] // Linux and OSX do not support some of these
```
Comment (stephentoub) [inline src/System.Net.NetworkInformation/tests/FunctionalTests/IPInterfacePropertiesTest.cs]: For those it doesn't support, can we add Unix-specific tests that invoke the property and verify it throws PlatformNotSupportedException?

**Comment by stephentoub** [comment]
A few comments, but otherwise LGTM.

File: src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Linux.cs (hunk 1 above)
Comment (mellinoe) [inline src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/NetworkAddressChange.Linux.cs]: I added an assertion into CreateSocket() that checks the current s_socket == 0 before creating a new one.

File: src/System.Net.NetworkInformation/tests/FunctionalTests/IPInterfacePropertiesTest.cs (hunk 2 above)
Comment (mellinoe) [inline src/System.Net.NetworkInformation/tests/FunctionalTests/IPInterfacePropertiesTest.cs]: I'll be adding more unix-specific tests separately, and I will probably split these into more granular tests then.

**Comment by mellinoe** [comment]
Thanks, Stephen, merging in as CI is green.

**Pull request merged**